File: bumperbot_cpp_examples/src/simple_turtlesim_kinematics.cpp
#include "bumperbot_cpp_examples/simple_turtlesim_kinematics.hpp"
#include <cmath>

using std::placeholders::_1;

SimpleTurtlesimKinematics::SimpleTurtlesimKinematics(const std::string &name) : Node(name){

    turtle1_pose_sub_ = create_subscription<turtlesim::msg::Pose> ("/turtle1/pose", 10, 
                                                                   std::bind(&SimpleTurtlesimKinematics::turtle1PoseCallback, this, _1));
    turtle2_pose_sub_ = create_subscription<turtlesim::msg::Pose> ("/turtle2/pose", 10, 
                                                                   std::bind(&SimpleTurtlesimKinematics::turtle2PoseCallback, this, _1));

    RCLCPP_INFO_STREAM(get_logger(), "using the rclcpp!");
 
}


void SimpleTurtlesimKinematics::turtle1PoseCallback(const turtlesim::msg::Pose & pose){
    last_turtle1_pose_ = pose;
    
}

void SimpleTurtlesimKinematics::turtle2PoseCallback(const turtlesim::msg::Pose & pose){
    last_turtle2_pose_ = pose;

    float Tx = last_turtle2_pose_.x - last_turtle1_pose_.x;
    float Ty = last_turtle2_pose_.y - last_turtle1_pose_.y;

    // Computation of the rotation matrix
    float theta_rad = last_turtle2_pose_.theta - last_turtle1_pose_.theta;
    float theta_deg = 180*theta_rad / 3.1416;

    RCLCPP_INFO_STREAM(get_logger(), "\nTranslation Vector trutle2 -> turtle1 \n" <<
                                     "Tx: " << Tx << "\n" <<
                                     "Ty: " << Ty << "\n" <<
                                     "Rotation Matrix turtle2 -> turtle1 \n" <<
                                     "theta_rad: " << theta_rad << "\n" <<
                                     "theta_deg: " << theta_deg << "\n" <<
                                     "|" << cos(theta_rad) << "      " << -sin(theta_rad) << "|\n" <<
                                     "|" << sin(theta_rad) << "      " << cos(theta_rad) << "|" );
} 

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTurtlesimKinematics>("simple_turtlesim_kinematics");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_py_examples/bumperbot_py_examples/simple_turtlesim_kinematics.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose  
import math

class SimpleTurtlesimKinematics(Node):
    def __init__(self):
        super().__init__("simple_turtlesim_kinematics")

        self.turtle1_pos_sub_ = self.create_subscription(Pose, "/turtle1/pose", self.turtle1PoseCallback, 10)
        self.turtle2_pos_sub_ = self.create_subscription(Pose, "/turtle2/pose", self.turtle2PoseCallback, 10)

        self.last_turtle1_pose_ = Pose()
        self.last_turtle2_pose_ = Pose()

    def turtle1PoseCallback(self, msg):
        self.last_turtle1_pose_ = msg

    def turtle2PoseCallback(self, msg):
        self.last_turtle2_pose_ = msg

        Tx = self.last_turtle2_pose_.x - self.last_turtle1_pose_.x
        Ty = self.last_turtle2_pose_.y - self.last_turtle1_pose_.y

        # Calculation of the rotation matrix
        # 1.- Calcultate the relative orientation
        theta_rad = self.last_turtle2_pose_.theta - self.last_turtle1_pose_.theta
        theta_deg = 180 * theta_rad / 3.1416

        self.get_logger().info("""\n 
                               Translation Vector turtle1 -> turtle2 \n
                               Tx: %f \n
                               Ty: %f \n 
                               Rotation Matrix turtle1 -> turtle2 \n
                               theta(rad): %f
                               theta(deg): %f \n
                               |R11      R12| : |%f       %f|\n
                               |R21      R22| : |%f       %f|\n""" % (Tx, Ty, theta_rad, theta_deg, math.cos(theta_rad), -math.sin(theta_rad), 
                                                                      math.sin(theta_rad), math.cos(theta_rad)))
        
def main(args=None):
    rclpy.init()
    node = SimpleTurtlesimKinematics()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
    

if __name__ == "__main__":
    main()